src/core/mempool.rs

```diff
@@ -1360,16 +1640,231 @@ impl MemPoolDB {
+        let args: &[&dyn ToSql] = &[&u64_to_sql(min_height)?, &u64_to_sql(max_height)?];
+        query_rows(&self.conn(), sql, args)
+    }
+
+    /// Get the transaction tag list that represents the set of recent transactions we have.
+    /// Generate them with our node-local seed so that our txtag list is different from anyone
+    /// else's, w.h.p.
+    pub fn get_txtags(&self, seed: &[u8]) -> Result<Vec<TxTag>, db_error> {
+        self.get_bloom_txids().map(|txid_list| {
+            txid_list
+                .iter()
+                .map(|txid| TxTag::from_seed_and_txid(seed, txid))
+                .collect()
+        })
+    }
+
+    /// How many recent transactions are there -- i.e. within BLOOM_COUNTER_DEPTH block heights of
+    /// the chain tip?
+    pub fn get_num_recent_txs(conn: &DBConn) -> Result<u64, db_error> {
+        let max_height = match MemPoolDB::get_max_height(conn)? {
+            Some(h) => h,
+            None => {
+                // mempool is empty
+                return Ok(0);
+            }
+        };
+        let min_height = max_height.saturating_sub(BLOOM_COUNTER_DEPTH as u64);
+        let sql = "SELECT COUNT(txid) FROM mempool WHERE height > ?1 AND height <= ?2";
+        let args: &[&dyn ToSql] = &[&u64_to_sql(min_height)?, &u64_to_sql(max_height)?];
+        query_int(conn, sql, args).map(|cnt| cnt as u64)
+    }
+
+    /// Make a mempool sync request.
+    /// If sufficiently sparse, use a MemPoolSyncData::TxTags variant
+    /// Otherwise, use a MemPoolSyncData::BloomFilter variant
+    /// If force_bloom_filter is true, then always make a bloom filter.  The reason for doin this
+    /// is that it's faster to do this than making a txtag list, even though the bloom filter is a
+    /// larger data structure.
+    pub fn make_mempool_sync_data(&self) -> Result<MemPoolSyncData, db_error> {
+        let num_tags = MemPoolDB::get_num_recent_txs(self.conn())?;
+        if num_tags < self.max_tx_tags.into() {
+            let seed = self.bloom_counter.get_seed().clone();
+            let tags = self.get_txtags(&seed)?;
+            Ok(MemPoolSyncData::TxTags(seed, tags))
+        } else {
+            Ok(MemPoolSyncData::BloomFilter(self.get_txid_bloom_filter()?))
+        }
+    }
+
+    /// Get the next batch of transactions from our mempool that are *not* represented in the given
+    /// MemPoolSyncData.  Transactions are ordered lexicographically by randomized_txids.hashed_txid, since this allows us
+    /// to use the txid as a cursor while ensuring that each node returns txids in a deterministic random order
+    /// (so if some nodes are configured to return fewer than MAX_BLOOM_COUNTER_TXS transactions,
+    /// a requesting node will still have a good chance of getting something useful).
+    pub fn find_next_missing_transactions(
+        &self,
+        data: &MemPoolSyncData,
+        height: u64,
+        last_txid: &Txid,
```

Comment:
Can you name this so that we know that this isn't an actual txid, rather it's a randomized txid.

Reply:
Yup, done